### PR TITLE
Added support for nested fields in createCollectionItem

### PR DESCRIPTION
**Pull request by DigitalGoldfish**
This is basic work for fixing issues #17 & #10 (repeater & set support)

I refactored the code for createCollectionItem so that the part that concerns the handling of field values is in it's own function. In case of repeater & set fields this new function is then called recursively for their child fields so that at the end all field values of a collection entry are in the collection item.

The field format is backwards compatible with the previous format, only now the field value might contain another field too; since this should not cause any conflicts the GraphQL type inference should always and some additional values will show in the GraphQL Tree. 
At least in theory, practice might show that I overlooked something ...

There is one change compared to the previous implementation. Before all field properties were spread into the itemField Object `...collectionFields[collectionFieldName],` and later on most of this properties are deleted again. As I didn't see any reason for that I elected to only pick `type` from the field configuration and add the value. If there was in fact some reason for that let me know and I will revert it.

Each field therefore only looks like this and has no further properties: 
```
{
  type: "text". // or image, wysiwyg, set, repeater ...
  value: "some value" // scalar, object or array value
}
```
There is one exception however - a repeater field configured with the `fields` option will also have an additional attribute `valueType ` for convenience
```
{
  type: "text". // or image, wysiwyg, set, repeater ...
  value: "some value" // scalar, object or array value
  valueType: "fields object name"
}
```

This is only the first step in enabling support for nested field, the next steps are:
- Add support for nested fields to normalizeCollections* functions
- Add support for nested fields to CollectionItemNodeFactory (link*FieldsTo*Nodes)

Since this isn't the easiest change and affects a lot of the existing codebase I decided to create this pull request early so that you can review what is done so far and give feedback early.

### CODE REVIEW

**Comment by DigitalGoldfish** [comment]
> This is only the first step in enabling support for nested field, the next steps are:
> 
> Add support for nested fields to normalizeCollections* functions
> Add support for nested fields to CollectionItemNodeFactory (linkFieldsToNodes)

This steps are now implemented too and at least images seem to be handled correctly for the cases I can think off. I don't use assets & markdown fields a lot so I don't have a lot of test content for these but since they are handled identically I - of course, edge cases with bugs are probably inevitable. :(

@WhippetsAintDogs Would love it if you can review this soon

**Comment by WhippetsAintDogs** [comment]
@DigitalGoldfish, It will be reviewed by the end of the day. Looks really neat so far ! :)

**Comment by DigitalGoldfish** [comment]
There's one problem with the nested field implementation - layout fields are not supported as nested fields. This is because in setFieldsOnGraphQLNodeType it is not possible to modify existing fields and therefore only root fields can be set. At least that's my understanding so far.
One possible way around this limitation is to define Layout fields as separate nodes that are then linked to the respective fields.

File: src/CockpitService.js [hunk 1]
```diff
@@ -243,43 +227,80 @@ const createCollectionItem = (
...
+    if (collectionFieldType === 'repeater') {
+      const repeaterFieldOptions = collectionFieldConfiguration.options || {};
+      if (typeof repeaterFieldOptions.field !== 'undefined') {
+        itemField.value = collectionFieldValue.map(repeaterEntry => (
+          createCollectionField(repeaterEntry.value, repeaterFieldOptions.field)
+        ));
+      } else if (repeaterFieldOptions.fields !== 'undefined') {
+        itemField.value = collectionFieldValue.map(repeaterEntry => (
+          repeaterFieldOptions.fields.reduce((accumulator, currentFieldConfiguration) => {
+            if (currentFieldConfiguration.name === repeaterEntry.field.name) {
```
Comment (WhippetsAintDogs) [inline src/CockpitService.js]: `currentFieldConfiguration.name` and `repeaterEntry.field.name` will always be `undefined` since the property `name` doesn't exist on this object.  Maybe, you wanted to use the `label` property ?

Comment (WhippetsAintDogs) [reply src/CockpitService.js]: Wow, I see why you were mistaken. Cockpit's set uses the 'name' property and Cockpit's repeaters, the 'label' property. Kudos to them for the consistency :stuck_out_tongue_closed_eyes: !

**Comment by WhippetsAintDogs** [comment]
Finally, I've tried the sets and they work well ! Good job @DigitalGoldfish ! :smiley:

**Comment by WhippetsAintDogs** [comment]
Last request, could you add a section documenting the repeaters/sets in the Readme ? (I can do it at the same time I'll publish the package on NPM if you are to busy :) )

File: src/CockpitService.js (hunk 1 above)
Comment (DigitalGoldfish) [reply src/CockpitService.js]: I can see why I didn't spot that problem - i habitually assign the "name" attribute to all my fields in the repeater as I am used to that from the set fields configuration. 
Also almost all of the field types have a name and and a label with the label being used for the "nice, readable name" shown in the GUI while the name is used for the json attribute name. 

To resolve this particular situation I guess I would need to use either the name or the label (depending if the name is set or not) with the label being slugified so that it can be used without problems in GraphQL.

Comment (WhippetsAintDogs) [reply src/CockpitService.js]: That's fine to me.

File: src/helpers.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+      item[fieldName] && types.includes(item[fieldName].type)
+    )
+    .map(fieldName => (
+      item[fieldName]
+    ));
+
+  // process fields nested in set
+  Object.keys(item)
+    .filter(
+      fieldName => !!item[fieldName] && item[fieldName].type === 'set'
```
[the remark this answers is not in the document]
Comment (DigitalGoldfish) [reply src/helpers.js]: It's not necessary, I removed it.

**Comment by DigitalGoldfish** [comment]
> Last request, could you add a section documenting the repeaters/sets in the Readme ? (I can do it at the same time I'll publish the package on NPM if you are to busy :) )

I extended the documentation - but feel free to modify and extend as you think is best. I am not too happy with my explanation :)

File: src/CockpitService.js (hunk 1 above)
Comment (DigitalGoldfish) [reply src/CockpitService.js]: Readme is extended as well, feel free to rewrite though as english isn't my native language.
